src/operator/contrib/tensorrt.cc

```diff
@@ -0,0 +1,196 @@
+  }
+  auto graph = model_proto.graph();
+  auto first_input_type = graph.input(0).type().tensor_type();
+  auto dim_value = first_input_type.shape().dim(0).dim_value();
+  uint64_t batch_size = static_cast<uint64_t>(dim_value);
+  // Need to set up max workspace size based on device properties
+  nvinfer1::ICudaEngine* const trt_engine = ::onnx_to_tensorrt::onnxToTrtCtx(
+      node_param.serialized_onnx_graph, batch_size, 1 << 30);
+
+  LOG(INFO) << "TensorRT engine instantiated!!!";
+
+  tensorrt::NameToIdx_t output_map;
+  for (auto& el : node_param.output_map) {
+    output_map[el.first] = std::get<0>(el.second);
+  }
+  return GetPtrMapping(trt_engine, node_param.input_map, output_map);
+}
+
+void TRTParamParser(nnvm::NodeAttrs* attrs) {
+  using namespace mshadow;
+
+  TRTParam param_;
+
+  try {
+    param_.Init(attrs->dict);
+    common::Deserialize(&param_.input_map, param_.serialized_input_map);
+    common::Deserialize(&param_.output_map, param_.serialized_output_map);
+    param_.onnx_pb_graph.ParseFromString(param_.serialized_onnx_graph);
+  } catch (const dmlc::ParamError& e) {
+    std::ostringstream os;
+    os << e.what();
+    os << ", in operator " << attrs->op->name << "("
+       << "name=\"" << attrs->name << "\"";
+    for (const auto& k : attrs->dict) {
+      os << ", " << k.first << "=\"" << k.second << "\"";
+    }
+    os << ")";
+    throw dmlc::ParamError(os.str());
+  }
+
+  attrs->parsed = std::move(param_);
+}
+
+template <>
+void TRTCompute<cpu>(const OpStatePtr& state, const OpContext& ctx,
+                     const std::vector<TBlob>& inputs,
+                     const std::vector<OpReqType>& req,
+                     const std::vector<TBlob>& outputs) {
+  LOG(FATAL) << "TRTCompute not implemented on the CPU";
+}
+
+inline bool TRTInferShape(const NodeAttrs& attrs, std::vector<TShape>* in_shape,
+                          std::vector<TShape>* out_shape) {
+  const auto node_param = nnvm::get<TRTParam>(attrs.parsed);
+  for (auto& el : node_param.output_map) {
+    (*out_shape)[std::get<0>(el.second)] = std::get<1>(el.second);
+  }
+  return true;
+}
+
+inline bool TRTInferStorageType(const NodeAttrs& attrs, const int dev_mask,
+                                DispatchMode* dispatch_mode,
+                                std::vector<int>* in_storage_type,
+                                std::vector<int>* out_storage_type) {
+  return storage_type_assign(out_storage_type, mxnet::kDefaultStorage,
+                             dispatch_mode, DispatchMode::kFCompute);
+}
+
+inline bool TRTInferType(const NodeAttrs& attrs, std::vector<int>* in_dtype,
+                         std::vector<int>* out_dtype) {
+  const auto node_param = nnvm::get<TRTParam>(attrs.parsed);
+  for (auto& el : node_param.output_map) {
+    (*out_dtype)[std::get<0>(el.second)] = std::get<3>(el.second);
+  }
+  return true;
+}
+
+inline std::vector<std::string> TRTListInputNames(const NodeAttrs& attrs) {
```

Comment:
Does this keep the order of inputs consistent with their corresponding nodes' topological order in the original graph?

Reply:
no the order of input is random (I iterate through an nnvm::NodeEntryMap)

Reply:
If the input is random, would this mismatch the original user input data order? For example, if op A is TRT compatible and it has multiple inputs whose order is A.ListInputNames(), after you make a TRT subgraph for A, the input order of the whole graph may be different from before. When users call on the executor of the TRT graph for inference, their data inputs still keep the original order while the TRT graph is expecting another order. Would this cause mismatching problem?

Reply:
@reminisce I think that the user sees only the original symbol, and not the optimized graph after the rewrite, because the optimization happens at simple_bind time, but the executor gets a copy of the original symbol, not the optimized one. So, from the user's perspective, nothing changes. Also, the weight matrices are fed in via the shared_buffer, so their order also doesn't matter, since shared_buffer is a dict in Python and a std::unordered_map inm C++. To me the whole graph rewrite is enapsulated and the user doesn't see it at all. As for inputs to the graph (not weights), they are set via the executor's arg dict, which again is unordered (Python dict). Am I missing something?

Reply:
I see. This only works for using symbol API directly. For more common use cases, where Module API is adopted, this would not work since the order of inputs may be different after partitioning, right?

Reply:
@reminisce It's a good point regarding the Module. Unfortunately this initial PR does not support Module, partly because it relies on Symbol's `simple_bind` method to provide the shared buffer, which is necessary because TensorRT stores copies of the weights. Binding from Module hides away the access to the weights from the user, and the user providing the weights during binding makes this approach work not only with symbolic models, but also Gluon models converted to symbolic from the pre-trained GluonCV model zoo. The concern over Module is valid, and the support for it would probably require some eventual re-working on the integration, but I though it would be good to at least get this initial integration out there, and then broaden the scope to the more common Module approach.